src/code/z_sram.c

```diff
@@ -1,25 +1,884 @@
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_sram/func_800A82C8.s")
+typedef struct {
+    /* 0x00 */ char newf[6]; // string "ZELDAZ"
+    /* 0x06 */ s16 deaths;
+    /* 0x08 */ char playerName[8];
+    /* 0x10 */ s16 n64ddFlag;
+    /* 0x12 */ s16 healthCapacity; // "max_life"
+    /* 0x14 */ s16 health;         // "now_life"
+    /* 0x16 */ s8 magicLevel;
+    /* 0x17 */ s8 magic;
+    /* 0x18 */ s16 rupees;
+    /* 0x1A */ u16 swordHealth;
+    /* 0x1C */ u16 naviTimer;
+    /* 0x1E */ u8 magicAcquired;
+    /* 0x1F */ u8 unk_1F;
+    /* 0x20 */ u8 doubleMagic;
+    /* 0x21 */ u8 doubleDefense;
+    /* 0x22 */ u8 bgsFlag;
+    /* 0x23 */ u8 ocarinaGameReward;
+    /* 0x24 */ ItemEquips childEquips;
+    /* 0x2E */ ItemEquips adultEquips;
+    /* 0x38 */ u32 unk_38; // this may be incorrect, currently used for alignement
+    /* 0x3C */ char unk_3C[0x0E];
+    /* 0x4A */ s16 savedSceneNum;
+} SavePlayerData; // size = 0x4C
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_sram/func_800A8484.s")
+typedef struct {
+    /* 0x0000 */ SavePlayerData playerData; // "S_Private" substruct name
+    /* 0x004C */ ItemEquips equips;
+    /* 0x0058 */ Inventory inventory;
+    /* 0x00B8 */ SavedSceneFlags sceneFlags[124];
+    /* 0x0E48 */ FaroresWindData fw;
+    /* 0x0E70 */ char unk_E70[0x10];
+    /* 0x0E80 */ s32 gsFlags[6];
+    /* 0x0E98 */ char unk_E98[0x10];
+    /* 0x0EA8 */ s32 horseRaceRecord;
+    /* 0x0EAC */ char unk_EAC[0x0C];
+    /* 0x0EB8 */ u16 eventChkInf[14]; // "event_chk_inf"
+    /* 0x0ED4 */ u16 itemGetInf[4];   // "item_get_inf"
+    /* 0x0EDC */ u16 infTable[30];    // "inf_table"
+    /* 0x0F18 */ char unk_F18[0x04];
+    /* 0x0F1C */ u32 worldMapAreaData; // "area_arrival"
+    /* 0x0F20 */ char unk_F20[0x4];
+    /* 0x0F24 */ u8 scarecrowCustomSongSet;
+    /* 0x0F25 */ u8 scarecrowCustomSong[0x360];
+    /* 0x1285 */ char unk_1285[0x24];
+    /* 0x12A9 */ u8 scarecrowSpawnSongSet;
+    /* 0x12AA */ u8 scarecrowSpawnSong[0x80];
+    /* 0x132A */ char unk_132A[0x02];
+    /* 0x132C */ HorseData horseData;
+    /* 0x1336 */ u16 checksum; // "check_sum"
+} SaveInfo;                    // size = 0x1338
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_sram/func_800A88D4.s")
+typedef struct {
+    /* 0x00 */ s32 entranceIndex;
+    /* 0x04 */ s32 linkAge; // 0: Adult; 1: Child
+    /* 0x08 */ s32 cutsceneIndex;
+    /* 0x0C */ u16 dayTime; // "zelda_time"
+    /* 0x10 */ s32 nightFlag;
+    /* 0x14 */ s32 numDays;
+    /* 0x18 */ s32 unk_18;    // increments with numDays, gets reset by goron for bgs and one other use
+    /* 0x1C */ SaveInfo info; // "information"
+} Save;                       // size = 0x1354
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_sram/func_800A8A20.s")
+#define SAVE_PLAYER_DATA (*((SavePlayerData*)&gSaveContext.newf))
+#define SAVE_INFO (*((SaveInfo*)&gSaveContext.newf))
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_sram/func_800A9258.s")
+#define FIELD_SIZE(type, field) sizeof(((type*)0)->field)
```

Comment:
Not sure I get the need for this macro, why not just use sizeof on the type instead?

Reply:
@Random06457

Reply:
This was just because I initially made unnamed structs within structs but now that everything is split up properly in different structs, there's no need to keep this.

Reply:
```
#define SAVE_SIZE sizeof(Save)
#define INFO_SIZE sizeof(SaveInfo)
```
these do indeed work fine. probably no use for a define at this point right?

Separate but related questions. The Slot size is basically sizeof SaveContext plus some additional padding to round it up. Do we want to define it like `#define SLOT_SIZE sizeof(SaveContext) + 0x28` or just hardcode the number?
Similarly, SRAM size is 6 slots (3 main, 3 backup) plus some extra padding. So do we want to do 
`#define SRAM_SIZE (SLOT_SIZE * 6) + 0x620` ?

Reply:
The 0x8000 bytes that make SRAM_SIZE is hardware dependent so it definitely shouldn't be relative to the size of slots. 
As for SLOT_SIZE, I think it's better to have it as `sizeof(SaveContext) + 0x28`, for now at least.

Reply:
alright, sounds good.
I know that file select needs SRAM_SIZE, and it may need slot and checksum size as well. Where would be a good place for them to live? save.h or does it not fit there

Reply:
`z64.h` seems fine to me for the time being.